src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -644,11 +673,11 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
 
                             // Also compute the name of the union type, eg. `Foo` so we
                             // can add a helpful note with it.
-                            let ty = base.ty(self.mir, self.infcx.tcx).ty;
+                            let ty = Place::ty_from(&base, &proj_base, self.mir, self.infcx.tcx).ty;
```

Comment:
add a FIXME to ensure we don't compute the type twice (and instead have `is_union` return it)